chapter_appendix-mathematics-for-deep-learning/multivariable-calculus_vn.md

```diff
@@ -793,30 +793,33 @@ The fact that we compute derivatives from $f$ back towards the inputs rather tha
 Note that there are two steps:
 -->
 
-*dịch đoạn phía trên*
+Việc tính đạo hàm từ $f$ trở ngược về đầu vào hơn là từ đầu vào đến đầu ra (như chúng ta đã thực hiện ở đoạn code ở trên) là lý do cho cái tên *lan truyền ngược* (*backpropagation*) của thuật toán.
+Có hai bước:
 
 
 <!--
 1. Compute the value of the function, and the single step partials from front to back. While not done above, this can be combined into a single *forward pass*.
 2. Compute the gradient of $f$ from back to front.  We call this the *backwards pass*.
 -->
 
-*dịch đoạn phía trên*
+1. Tính giá trị của hàm và đạo hàm riêng theo từng bước đơn lẻ từ đầu đến cuối. Mặc dù chưa được thực hiện ở trên, việc này có thể kết hợp vào *lượt truyền xuôi*.
```

Comment:
Ai giải thích cái này giùm mình được không? Mình không hiểu lắm

Reply:
Theo mình hiểu như thế này: Đoạn code ngay trước đó minh hoạ cho việc tính toán forward và backward, trong đó họ tính xong một lượt forward từ đầu vào đến đầu ra rồi mới thực hiện tiinhs đạo hàm riêng qua mỗi bước từ đầu ra trở ngược về đầu vào như thế này:
```
# Compute the value of the function from inputs to outputs
w, x, y, z = -1, 0, -2, 1
a, b = (w + x + y + z)**2, (w + x - y - z)**2
u, v = (a + b)**2, (a - b)**2
f = (u + v)**2
print(f'f at {w}, {x}, {y}, {z} is {f}')

# Compute the derivative using the decomposition above
# First compute the single step partials
df_du, df_dv = 2*(u + v), 2*(u + v)
du_da, du_db, dv_da, dv_db = 2*(a + b), 2*(a + b), 2*(a - b), -2*(a - b)
da_dw, db_dw = 2*(w + x + y + z), 2*(w + x - y - z)
da_dx, db_dx = 2*(w + x + y + z), 2*(w + x - y - z)
da_dy, db_dy = 2*(w + x + y + z), -2*(w + x - y - z)
da_dz, db_dz = 2*(w + x + y + z), -2*(w + x - y - z)
```
**"1. Compute the value of the function, and the single step partials from front to back. While not done above, this can be combined into a single *forward pass*."** (*Tính giá trị của hàm và đạo hàm riêng theo từng bước đơn lẻ từ đầu đến cuối. Mặc dù chưa được thực hiện ở trên, việc này có thể kết hợp vào *lượt truyền xuôi*.) có nghĩa là với mỗi bước đơn forward, bạn tính luôn đạo hàm riêng cho từng biến trong bước đó mà không cần phải forward xong xuôi từ đầu đến cuối rồi mới tính đạo hàm riêng. Tức có nghĩa là họ sẽ tính toán như này:
```
# Compute the value of the function, and the single step partials from front to back.
w, x, y, z = -1, 0, -2, 1
## Single step 1
a, b = (w + x + y + z)**2, (w + x - y - z)**2 # forward
da_dw, db_dw = 2*(w + x + y + z), 2*(w + x - y - z) # partial
da_dx, db_dx = 2*(w + x + y + z), 2*(w + x - y - z) # partial
da_dy, db_dy = 2*(w + x + y + z), -2*(w + x - y - z) # partial
da_dz, db_dz = 2*(w + x + y + z), -2*(w + x - y - z) # partial
## Single step 2
u, v = (a + b)**2, (a - b)**2 # forward
du_da, du_db, dv_da, dv_db = 2*(a + b), 2*(a + b), 2*(a - b), -2*(a - b) #partial
## Single step 3
f = (u + v)**2 #forward 
df_du, df_dv = 2*(u + v), 2*(u + v) #partial
```
Nếu câu đó tối nghĩa mình đề xuất dịch lại thế này:
`Tính giá trị của hàm và đạo hàm riêng theo từng bước đơn lẻ từ đầu đến cuối. Mặc dù chưa được thực hiện ở trên, việc tính đạo hàm riêng có thể kết hợp vào *lượt truyền xuôi* đơn lẻ.

Reply:
@nguyenhoa93 mình thấy cách phía dưới là tính dần đạo hàm riêng qua mỗi layer, cách trên là truyền xuôi qua hết rồi tính 1 thể. Bản chất mình nghĩ như nhau, vẫn là có thể tích hợp trong 1 lượt truyền xuôi (combined into a single forward pass). 
Mình cũng không hiểu `While not done above` đoạn này lắm.

Reply:
Thanks @nguyenhoa93 , mình có thể hiểu là cách thứ nhất thì muốn tính luôn d_f/d_x còn cách hai thì forward đến tầng nào thì tính d_o/d_x trước (đạo hàm của đầu ra theo đầu vào) sau đó trong lượt truyền ngược mới tính d_f/d_x

Reply:
@cuongvng mình nghĩ `while not done above` tác giả ám chỉ ở đoạn code phía trên, khi họ tách lượt truyền xuôi ra thành: tính toán giá trị đầu ra và tính đạo hàm riêng đơn bước

Reply:
```suggestion
1. Tính giá trị của hàm và đạo hàm riêng theo từng bước đơn lẻ từ đầu đến cuối. Mặc dù không được thực hiện ở trên, 
 hai việc này có thể được kết hợp vào một *lượt truyền xuôi* duy nhất.
```